### ISSUE TITLE
Live push logs

### ISSUE THREAD

**Issue by p3lim** [timbrel/GitSavvy#1021]
When pushing to a remote that responds with some build information (such as Heroku), it will stream the response to the client.
However, when pushing with GitSavvy, I only get the complete log _once the push has finished_, and not streamed like in a terminal.

Not sure if there are limits with the ST API, but support for this would be nice.
(The `[Done in XXs]` at the end is more than sufficient to indicate the process completion).

ST: 3176
Git: 2.19.0
OS: Arch, kernel 4.18.5

**Comment by randy3k** [comment]
Check the setting `live_panel_output`.

**Comment by randy3k** [comment]
@p3lim I am going to close this issue now. Please get us know if `live_panel_output` is not what you are looking for.

**Comment by p3lim** [comment]
It's probably what I would want, but it doesn't seem to work like it should.
What I expect is that the output is streamed to the panel as it arrives, but it's only streamed once it's done (as far I could see when I tested it).

**Comment by randy3k** [comment]
The output should stream to the panel (if I remember correctly).

**Comment by p3lim** [comment]
What I'm experiencing is the output being streamed, but only once the process is finished, not as it's running.

**Comment by randy3k** [comment]
I am not very familiar with Heroku, do you just `git push` to have the result streamed to the terminal?

**Comment by randy3k** [comment]
Did you set
```
    "show_panel_for": [
        "push"
    ],
```

**Comment by p3lim** [comment]
Yes, that option is set.

When I push to Heroku it triggers a deployment process on their side, which the log for is streamed as part of the push.

**Comment by randy3k** [comment]
Let's reopen this for now. Though I still don't understand why the setting `live_panel_output` does not work.

**Comment by asfaltboy** [comment]
Live panel output isn't showing for me when pushing to Heroku either. Seems the issue is that while we're communicating with the background threads, the ST loop isn't executing our "async panel commands" (i.e commands we ran with `set_timeout_async`).

I was sort of able to get around by updating `panel` & `panel_append` to support running with `run_async=False`. This work can be seen in https://github.com/divmain/GitSavvy/tree/1021-live-panel-output-fix

But, this causes all other commands to freeze while we (e.g) spend a long time deploying to Heroku. I think we might be able to resolve this by some sort of sleep call, I'll try and hack on this a bit more.

**Comment by asfaltboy** [comment]
Oops, I believe I got it reverse, I think it is actually the main thread (where we run our push command + wait on process) that does not relinquish it's control to other GitSavvy commands (like calling panel). By making the calls sync, we avoid the panel commands to be stuck in the loop's queue!

I guess this would always be an issue, regardless of the panel, so my POC branch should be ok to merge, at least to solve the live_panel not showing issue.

To solve the manner of long running blocking git commands, we may want to look into running specific (all?) commands *in sync mode*, but _in a background thread_, and loop/sleep until they are done. This way we should be able to call multiple commands and keep reference to all child threads that are still running. Perhaps we can start "gently" by implementing such a feature as opt-in for specified command/s, and play around with it?

**Comment by randy3k** [comment]
TLDR, it is because Sublime Text has only one background thread (set_timeout_async)?

**Comment by kaste** [comment]
This has been fixed
